[ondilo] Initial contribution (#18914)
openhab/openhab-addons: bundles/org.openhab.binding.ondilo/README.md

# Ondilo Binding

This binding integrates Ondilo ICO pool monitoring devices with openHAB, allowing you to monitor and automate your pool environment using openHAB’s rules and UI.

## Supported Things

`bridge:` Represents your Ondilo account (OAuth2)

> 

Reviewer: ```suggestion

```

Author: removing the line with the bridge was the intention here?

`ondilo:` Represents an individual Ondilo ICO device

Ondilo ICO Pool as well as Spa devices are supported.
Chlor as well as salt water.

## Discovery

Ondilo ICOs are discovered automatically after the bridge is authorized and online.
Each Ondilo ICO will appear as a new Thing in the inbox.

## Thing Configuration

### `bridge` Thing Configuration

- **url**: The URL of the openHAB instance. Required for the redirect during OAuth2 flow (e.g. `http://localhost:8080`)

> - **refreshInterval**: Polling interval of the bridge (i.e. config changes) (default: `1800 s`)

Reviewer: ```suggestion
- **refreshInterval**: Polling interval of the bridge (i.e. configuration changes) in seconds (default: `1800 s`)

```

Author: fixed

### `ondilo` Thing Configuration

- **id**: The Id of an Ondilo ICO device. Set via discovery service (e.g. `12345`)

> - **refreshInterval**: Polling interval of the device (i.e. measures) (default: `600 s`)

Reviewer: ```suggestion
- **refreshInterval**: Polling interval of the device (i.e. measures) in seconds (default: `600 s`)

```

Author: fixed

Ondilo ICO takes measures every hour.

> Higher polling will not increase the update interval.

Reviewer: Is it predicatable when the measure is taking place? In that case a cron scheduler or other mechanism could perform a poll just after the physicall device is updated. keeping the value in openHAB as up to date as possible.

Author: 1 measure every hour. A poll can be performed, i.e. 61 min after the last measure.
I think I will keep cyclic polling, but trying to sync the schedule with the predicted measure time.

Author: This is now implemented with the latest changes

The requests to the Ondilo Customer API are limited to the following per user quotas:

- 5 requests per second
- 30 requests per hour

`bridge` Thing performs 1 request per cycle - 2 per hour with default interval.

> 

Reviewer: ```suggestion

```

Author: fixed

> `ondilo` Thing performs 3 request per cycle - 18 per hour with default interval.

Reviewer: JFTR: when starting both bridge and ondilo thing (like on swtartup) are danger close to the limit 4/5 per second. If you have multiple things, you would exceed the rate limit?

Author: Using different initial delays reduces the risk.

2 Things would still cause 6 calls per second. I'm not sure, how strict the quota is and how much energy needs to be invested into this.

Author: Things has now an initial delay of 2 s while Bridge has 1 s.

There would be a bigger API request handler logic required, if users have several pools.
E.g. Poll Bridge in S0, Thing 1 in S1, Thing 2 in S2 ...
This could be implemented during the bigger Pool handler <=> Bridge Handler <=> ApiClient change that you are asking for.

|---------------------------|-------------------------|----------|--------|--------------------------------------------------------|
| temperature               | Number:Temperature      | false    | R      | Water temperature in the pool                          |
| ph                        | Number                  | false    | R      | pH value of the pool water                             |
| orp                       | Number:ElectricPotential| false    | R      | Oxidation-reduction potential (ORP)                    |
| salt                      | Number:Density          | false    | R      | Salt concentration in the pool (salt pools only)       |
| tds                       | Number:Density          | false    | R      | Total dissolved solids in the pool (chlor pools only ) |
| battery                   | Number:Dimensionless    | false    | R      | Battery level of the device                            |
| rssi                      | Number:Dimensionless    | false    | R      | Signal strength (RSSI)                                 |
| value-time                | DateTime                | true     | R      | Timestamp of the set of measures                       |

> | poll-update               | Switch                  | true     | R/W    | Poll status update from the cloud (get latest measures, not a trigger for new measures) |

Author: a manual poll

### Recommendations Channels

| Channel ID                | Type                    | Advanced | Access | Description                                            |
|---------------------------|-------------------------|----------|--------|--------------------------------------------------------|
| recommendation-id         | Number                  | true     | R      | Unique ID of the current recommendation                |
| recommendation-title      | String                  | false    | R      | Title of the current recommendation                    |
| recommendation-message    | String                  | false    | R      | Message of the current recommendation                  |

> | recommendation-created_at | String                  | true     | R      | Creation time of the current recommendation            |
> | recommendation-updated_at | String                  | true     | R      | Last update time of the current recommendation         |

Reviewer: ```suggestion
| recommendation-created-at | String                  | true     | R      | Creation time of the current recommendation            |
| recommendation-updated-at | String                  | true     | R      | Last update time of the current recommendation         |

```

Author: fixed

| recommendation-status     | String                  | false    | R      | Status of the current recommendation                   |
| recommendation-deadline   | String                  | true     | R      | Deadline of the current recommendation                 |

## Full Example

### Thing Configuration

```Java
Bridge ondilo:bridge:mybridge [ url="http://localhost:8080", refreshInterval=1800 ] {

>     Thing ondilo "<id_received_from_discovery>" [ mowerId="<id_received_from_discovery>" ] {

Author: you got me ;-) I copied allot from the automower binding :-o

When will you find time to review my open PR there?

>     Thing ondilo "<id_received_from_discovery>" [ mowerId="<id_received_from_discovery>" ] {

Author: fixed

    }
```

> ### Item Configuration

Author: The template supports the original version: https://github.com/openhab/openhab-core/blob/main/tools/archetype/binding/src/main/resources/archetype-resources/README.md

```java
Number:Temperature        Ondilo_Temperature  "Pool Temperature [%.1f %unit%]"  { channel="ondilo:ondilo:mybridge:myonilo:measure#temperature" }
Number                    Ondilo_pH           "Pool pH [%d]"                    { channel="ondilo:ondilo:mybridge:myonilo:measure#ph" }
Number:ElectricPotential  Ondilo_ORP          "Pool ORP [%.1f %unit%]"          { channel="ondilo:ondilo:mybridge:myonilo:measure#orp" }
Number:Density            Ondilo_Salt         "Pool Salt [%.0f %unit%]"         { channel="ondilo:ondilo:mybridge:myonilo:measure#salt" }
Number:Dimensionless      Ondilo_Battery      "Pool Battery [%d %]"             { channel="ondilo:ondilo:mybridge:myonilo:measure#battery" }
Number:Dimensionless      Ondilo_RSSI         "Pool RSSI [%.0f]"                { channel="ondilo:ondilo:mybridge:myonilo:measure#rssi" }

String                    Ondilo_RecTitle     "Recommendation Title [%s]"       { channel="ondilo:ondilo:mybridge:myonilo:recommendation#title" }
String                    Ondilo_RecMessage   "Recommendation Message [%s]"     { channel="ondilo:ondilo:mybridge:myonilo:recommendation#message" }

> String                    Ondilo_RecStatus    "Recommendation Status [%s]"      { channel="ondilo:ondilo:mybridge:myonilo:recommendation#status" }

Author: fixed

```

> ### Sitemap Configuration

Author: The template supports the original version: https://github.com/openhab/openhab-core/blob/main/tools/archetype/binding/src/main/resources/archetype-resources/README.md
